### PR TITLE
webgpu: Relax the condition of MatMulReduceProgram

### PR DESCRIPTION
**Pull request by qjia7**
PERF

To see the logs from the Cloud Build CI, please join either our [discussion](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs) or [announcement](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs-announce) mailing list.

---
This change is [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/6539)

### CODE REVIEW

**Comment by qjia7** [comment]
Tested on CFL
Before:
<html>
<body>


Conv2D | 0.42 | input0: 4D[1,1,2,2048]input1: 4D[1,1,2048,48] | 1,1,2,48 | MatMulSmallOutputSizeProgram: 0.415326
-- | -- | -- | -- | --
Conv2D | 0.41 | input0: 4D[1,1,2,2048]input1: 4D[1,1,2048,17] | 1,1,2,17 | MatMulSmallOutputSizeProgram: 0.41116
Conv2D | 0.41 | input0: 4D[1,1,2,2048]input1: 4D[1,1,2048,34] | 1,1,2,34 | MatMulSmallOutputSizeProgram: 0.41066
Conv2D | 0.41 | input0: 4D[1,1,2,2048]input1: 4D[1,1,2048,24] | 1,1,2,24 | MatMulSmallOutputSizeProgram: 0.40766
Conv2D | 0.40 | input0: 4D[1,1,2,2048]input1: 4D[1,1,2048,34] | 1,1,2,34 | MatMulSmallOutputSizeProgram: 0.396327
Conv2D | 0.40 | input0: 4D[1,1,2,2048]input1: 4D[1,1,2048,32] | 1,1,2,32 | MatMulSmallOutputSizeProgram: 0.395994
Conv2D | 0.40 | input0: 4D[1,1,2,2048]input1: 4D[1,1,2048,32] | 1,1,2,32 | MatMulSmallOutputSizeProgram: 0.395661


</body>
</html>
After:
<html>
<body>


Conv2D | 0.18 | input0: 4D[1,1,2,2048]input1: 4D[1,1,2048,48] | 1,1,2,48 | MatMulReduceProgram: 0.178747
-- | -- | -- | -- | --
Conv2D | 0.13 | input0: 4D[1,1,2,2048]input1: 4D[1,1,2048,34] | 1,1,2,34 | MatMulReduceProgram: 0.134581
Conv2D | 0.13 | input0: 4D[1,1,2,2048]input1: 4D[1,1,2048,34] | 1,1,2,34 | MatMulReduceProgram: 0.132581
Conv2D | 0.13 | input0: 4D[1,1,2,2048]input1: 4D[1,1,2048,32] | 1,1,2,32 | MatMulReduceProgram: 0.126498
Conv2D | 0.13 | input0: 4D[1,1,2,2048]input1: 4D[1,1,2048,32] | 1,1,2,32 | MatMulReduceProgram: 0.126082
Conv2D | 0.11 | input0: 4D[1,1,2,2048]input1: 4D[1,1,2048,24] | 1,1,2,24 | MatMulReduceProgram: 0.106582
Conv2D | 0.08 | input0: 4D[1,1,2,2048]input1: 4D[1,1,2048,17] | 1,1,2,17 | MatMulReduceProgram: 0.076499

</body>
</html>

**Review by gyagp — APPROVED** [review]
Good improvement. LGTM!

File: tfjs-backend-webgpu/src/kernels/BatchMatMul_impl.ts [hunk 1]
```diff
@@ -94,7 +94,7 @@ export function batchMatMulImpl({
                    (outerShapeA % 4 === 0 && transposeA)) &&
       outerShapeB % 4 === 0 && !transposeB;
   let program: WebGPUProgram;
-  if (outerShapeA * outerShapeB <= 32) {
```
Comment (xhcao) [inline tfjs-backend-webgpu/src/kernels/BatchMatMul_impl.ts:97]: From the point of global memory load, `MatMulReduceProgram` is similar as a naive matmul method, the count of global memory load is 2*M*K*N. For the other shared memory version of matmul, the count is M*K*N*(1/tileColSize + 1/tileRowSize). For small size matmul, GPU bandwidth will be enough and no bottleneck happens when using  `MatMulReduceProgram`.  I think that different hardwares have its own critical value, so we need to test on more hardwares and find a conservative value for the condition.
And [2, 2048, 64] is not friend to other matmul versions, but [8, 2048, 16] is very friend, so you should test the cases, and compare their performance.

Comment (qjia7) [reply tfjs-backend-webgpu/src/kernels/BatchMatMul_impl.ts:97]: Below is the test result with [8, 2048, 16]
```
CFL
Matmul[8, 2048, 16]: 0.133998 ms, MatMulReduceProgram: 0.133998
Matmul[8, 2048, 16]: 0.134915 ms, MatMulReduceProgram: 0.134915
Matmul[8, 2048, 16]: 0.266079 ms, MatMulSmallOutputSizeProgram: 0.266079
Matmul[8, 2048, 16]: 0.284662 ms, MatMulSmallOutputSizeProgram: 0.284662
TGL
Matmul[8, 2048, 16]: 0.056562 ms, MatMulReduceProgram: 0.056562
Matmul[8, 2048, 16]: 0.055417 ms, MatMulReduceProgram: 0.055417
Matmul[8, 2048, 16]: 0.281092 ms, MatMulSmallOutputSizeProgram: 0.281092
Matmul[8, 2048, 16]: 0.280988 ms, MatMulSmallOutputSizeProgram: 0.280988
```
You can see that `MatMulReduceProgram` is still faster than current default matmul pass.

**Review by xhcao — APPROVED** [review]
Thank you for the great performance data, LGTM

**Pull request merged**